### ISSUE TITLE
Limit recursion depth based on size/range instead of configuration value

### ISSUE THREAD

**Issue by bender2k14** [hedgehogqa/fsharp-hedgehog-experimental#37]
PR #21 added a configuration value called `RecursionDepth` to prevent stack overflows when generating a recursive type (c.f. https://github.com/hedgehogqa/fsharp-hedgehog-experimental/issues/19#issuecomment-359886908).

[FsCheck solves this problem](https://fscheck.github.io/FsCheck/TestData.html#main:~:text=However%2C%20a%20recursive%20generator%20like%20this,always%20use%20the%20size%20control%20mechanism) by relating the recursion depth to the size parameter.
> However, a recursive generator like this may fail to terminate with a `StackOverflowException`, or produce very large results. To avoid this, recursive generators should always use the size control mechanism

This seems like a better solution to me, but I am just getting seriously into property-based testing, so there are probably many things to consider of which I am unaware.

What are the tradeoffs for this configuration value vs utilizing the size parameter?

**Comment by cmeeren** [comment]
I have no idea; @moodmosaic?

**Comment by moodmosaic** [comment]
The _sized_ approach of FsCheck is more elegant since you can use the size parameter via `Gen.sized` to control the depth (instead of introducing another variable).

**Comment by moodmosaic** [comment]
Some thoughts around the `size`, in general: https://github.com/hedgehogqa/haskell-hedgehog/issues/38#issuecomment-293458877

**Comment by bender2k14** [comment]
That is what I expected.

@cmeeren, how do you feel about removing `RecursionDepth` and instead bounding recursive generation with the size parameter?

I am interested to implement this simply because I think it might be a good challenge for me.

**Comment by bender2k14** [comment]
> Some thoughts around the `size`, in general: [hedgehogqa/haskell-hedgehog#38 (comment)](https://github.com/hedgehogqa/haskell-hedgehog/issues/38#issuecomment-293458877)

Yes, I really like the `Range` concept.  I think it is very clear...after @cmeeren first explained to me in https://github.com/elmish/Elmish.WPF/pull/240#issuecomment-651989014 and then especially after watching the video @moodmosaic linked to in https://github.com/hedgehogqa/fsharp-hedgehog/issues/182#issuecomment-466997307.

**Comment by cmeeren** [comment]
> @cmeeren, how do you feel about removing `RecursionDepth` and instead bounding recursive generation with the size parameter?

I'm happy to use `size` to control the recursion depth, but perhaps we should have a `MaxRecursionDepth` so that users can ensure it doesn't recurse that much (too much recursion in auto-generation can substantially increase the test run time). What do you think?

**Comment by moodmosaic** [comment]
>so that users have some control

They’ll have that control through the size... What happens when `size > MaxRecursionDepth`?

**Comment by cmeeren** [comment]
> They’ll have that control through the size...

How so? Isn't `size` automatically incremented from 0 to 99 or something like that based on the test that is run?

**Comment by moodmosaic** [comment]
Yes, but you can interpret it (in this case) any way you want, e.g. halve it on each recursion (as FsCheck mentions in the docs).

**Comment by cmeeren** [comment]
Ok. (For reference, [here's the mentioned docs](https://fscheck.github.io/FsCheck/TestData.html#Generating-recursive-data-types).)

I won't be able to work on this, but I'm more than happy to take a PR that makes sure that auto-generating recursive types aren't unnecessarily slow, and gives the user some kind of control over the recursion depth. The readme should have an example of how to control it.
